playwright/_impl/_assertions.py

```diff
@@ -0,0 +1,539 @@
+# Copyright (c) Microsoft Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from typing import Any, List, Pattern, Union
+from urllib.parse import urljoin
+
+from playwright._impl._api_structures import ExpectedTextValue, FrameExpectOptions
+from playwright._impl._locator import Locator
+from playwright._impl._page import Page
+
+
+class AssertionsBase:
+    def __init__(self, locator: Locator, is_not: bool = False) -> None:
+        self._actual_locator = locator
+        self._loop = locator._loop
+        self._is_not = is_not
+
+    async def _expect_impl(
+        self,
+        expression: str,
+        expect_options: FrameExpectOptions,
+        expected: Any,
+        message: str,
+    ) -> None:
+        expect_options["isNot"] = self._is_not
+        if expect_options.get("timeout") is None:
+            expect_options["timeout"] = 5_000
+        if expect_options["isNot"]:
+            message = message.replace("expected to", "expected not to")
+        if "useInnerText" in expect_options and expect_options["useInnerText"] is None:
+            del expect_options["useInnerText"]
+        result = await self._actual_locator._expect(expression, expect_options)
+        if result["matches"] == self._is_not:
+            log = "".join(result.get("log", ""))
+            if log.strip():
+                log = "\nCall log:\n" + log
+            if expected is None:
+                raise AssertionError(f"{message} {log}")
+            raise AssertionError(f"{message} {log}")
+
+
+class PageAssertions(AssertionsBase):
+    def __init__(self, page: Page, is_not: bool = False) -> None:
+        super().__init__(page.locator(":root"), is_not)
+        self._actual_page = page
+
+    @property
+    def _not(self) -> "PageAssertions":
+        return PageAssertions(self._actual_page, not self._is_not)
+
+    async def to_have_title(
+        self, title_or_reg_exp: Union[Pattern, str], timeout: float = None
+    ) -> None:
+        expected_values = to_expected_text_values(
+            [title_or_reg_exp], normalize_white_space=True
+        )
+        await self._expect_impl(
+            "to.have.title",
+            FrameExpectOptions(expectedText=expected_values, timeout=timeout),
+            title_or_reg_exp,
+            "Page title expected to be",
```

Comment:
Could you please show an example of how the error looks like?

Reply:
https://gist.github.com/mxschmitt/1af1ee657a4237fafb13588b0324ceb3